File: Sources/Models/SmartBlock.swift
import Foundation

// MARK: - SmartBlock Type

/// Unified block type shared across chat, classroom, and course generation.
/// Forward-compatible: decodes unknown backend types to `.unknown` with fallback rendering.
public enum SmartBlockType: String, Codable, Sendable {
    case text           // subtypes: heading, paragraph, summary, callout, hook, revelation
    case code           // subtypes: snippet, playground, terminal
    case quiz           // subtypes: mcq, trueFalse, fillBlank, poll
    case flashcard      // subtypes: single, deck
    case dataViz        // subtypes: chart, graph, diagram, math, table
    case media          // subtypes: image, video, audio, animation
    case progress       // subtypes: checkpoint, celebration, divider, spacer, timeline
    case interactive    // subtypes: comparison, stepByStep, notes
    case masteryMap
    case unknown
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let raw = try container.decode(String.self)
        // Handle both camelCase ("dataViz") and snake_case ("data_viz") from backend
        if let direct = SmartBlockType(rawValue: raw) {
            self = direct
        } else {
            // Convert snake_case to camelCase for rawValue lookup
            let parts = raw.split(separator: "_")
            let camelCased = parts.enumerated().map { index, part in
                index == 0 ? String(part) : part.capitalized
            }.joined()
            self = SmartBlockType(rawValue: camelCased) ?? .unknown
        }
    }
}

// MARK: - SmartBlock Content

/// Type-safe content container — each type has its own payload.
/// Encoding/decoding is handled by the parent `SmartBlock`.
public enum SmartBlockContent: Sendable {
    case text(TextBlockPayload)
    case code(CodeBlockPayload)
    case quiz(QuizBlockPayload)
    case flashcard(FlashcardBlockPayload)
    case dataViz(DataVizBlockPayload)
    case media(MediaBlockPayload)
    case progress(ProgressBlockPayload)
    case interactive(InteractiveBlockPayload)
    case masteryMap(MasteryMapBlockPayload)
    case unknown(rawJSON: [String: AnyCodable])
}

// MARK: - SmartBlock

/// A single renderable content block with schema versioning.
public struct SmartBlock: Codable, Identifiable, Sendable {
    public let id: String
    public let schemaVersion: Int
    public let type: SmartBlockType
    public let subtype: String?
    public let content: SmartBlockContent
    public let metadata: [String: AnyCodable]?
    
    public init(
        id: String = UUID().uuidString,
        schemaVersion: Int = 1,
        type: SmartBlockType,
        subtype: String? = nil,
        content: SmartBlockContent,
        metadata: [String: AnyCodable]? = nil
    ) {
        self.id = id
        self.schemaVersion = schemaVersion
        self.type = type
        self.subtype = subtype
        self.content = content
        self.metadata = metadata
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(String.self, forKey: .id) ?? UUID().uuidString
        schemaVersion = try container.decodeIfPresent(Int.self, forKey: .schemaVersion) ?? 1
        type = try container.decode(SmartBlockType.self, forKey: .type)
        subtype = try container.decodeIfPresent(String.self, forKey: .subtype)
        metadata = try container.decodeIfPresent([String: AnyCodable].self, forKey: .metadata)
        
        // Decode content dict, then route to typed payload based on `type`
        let contentDict = try container.decode([String: AnyCodable].self, forKey: .content)
        let contentData = try JSONSerialization.data(
            withJSONObject: contentDict.mapValues { $0.value }
        )
        let dec = JSONDecoder()
        dec.keyDecodingStrategy = .convertFromSnakeCase
        
        switch type {
        case .text:
            content = .text(
                (try? dec.decode(TextBlockPayload.self, from: contentData))
                ?? TextBlockPayload(text: contentDict["text"]?.value as? String ?? "")
            )
        case .code:
            content = .code(
                (try? dec.decode(CodeBlockPayload.self, from: contentData))
                ?? CodeBlockPayload(language: "plain", code: contentDict["code"]?.value as? String ?? "")
            )
        case .quiz:
            content = .quiz(
                (try? dec.decode(QuizBlockPayload.self, from: contentData))
                ?? QuizBlockPayload(question: "", options: [], correctIndex: 0)
            )
        case .flashcard:
            content = .flashcard(
                (try? dec.decode(FlashcardBlockPayload.self, from: contentData))
                ?? FlashcardBlockPayload(front: "", back: "")
            )
        case .dataViz:
            content = .dataViz(
                (try? dec.decode(DataVizBlockPayload.self, from: contentData))
                ?? DataVizBlockPayload(format: "text", source: "")
            )
        case .media:
            content = .media(
                (try? dec.decode(MediaBlockPayload.self, from: contentData))
                ?? MediaBlockPayload(url: "", alt: nil)
            )
        case .progress:
            content = .progress(
                (try? dec.decode(ProgressBlockPayload.self, from: contentData))
                ?? ProgressBlockPayload(completed: 0, total: 1)
            )
        case .interactive:
            content = .interactive(
                (try? dec.decode(InteractiveBlockPayload.self, from: contentData))
                ?? InteractiveBlockPayload(items: [])
            )
        case .masteryMap:
            content = .masteryMap(
                (try? dec.decode(MasteryMapBlockPayload.self, from: contentData))
                ?? MasteryMapBlockPayload(title: "Course", nodes: [])
            )
        case .unknown:
            content = .unknown(rawJSON: contentDict)
        }
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(schemaVersion, forKey: .schemaVersion)
        try container.encode(type, forKey: .type)
        try container.encodeIfPresent(subtype, forKey: .subtype)
        try container.encodeIfPresent(metadata, forKey: .metadata)
        
        switch content {
        case .text(let p): try container.encode(p, forKey: .content)
        case .code(let p): try container.encode(p, forKey: .content)
        case .quiz(let p): try container.encode(p, forKey: .content)
        case .flashcard(let p): try container.encode(p, forKey: .content)
        case .dataViz(let p): try container.encode(p, forKey: .content)
        case .media(let p): try container.encode(p, forKey: .content)
        case .progress(let p): try container.encode(p, forKey: .content)
        case .interactive(let p): try container.encode(p, forKey: .content)
        case .masteryMap(let p): try container.encode(p, forKey: .content)
        case .unknown(let dict): try container.encode(dict, forKey: .content)
        }
    }
    
    private enum CodingKeys: String, CodingKey {
        case id, schemaVersion, type, subtype, content, metadata
    }
}

// MARK: - Content Payloads

public struct TextBlockPayload: Codable, Sendable {
    public let text: String
    public var style: String?  // heading, paragraph, summary, callout, hook, revelation
    
    public init(text: String, style: String? = nil) {
        self.text = text
        self.style = style
    }
}

public struct CodeBlockPayload: Codable, Sendable {
    public let language: String
    public let code: String
    public var isRunnable: Bool?
    
    public init(language: String, code: String, isRunnable: Bool? = nil) {
        self.language = language
        self.code = code
        self.isRunnable = isRunnable
    }
}

public struct QuizBlockPayload: Codable, Sendable {
    public let question: String
    public let options: [QuizOptionPayload]
    public let correctIndex: Int
    public var explanation: String?
    public var hint: String?
    
    public init(question: String, options: [QuizOptionPayload], correctIndex: Int, explanation: String? = nil, hint: String? = nil) {
        self.question = question
        self.options = options
        self.correctIndex = correctIndex
        self.explanation = explanation
        self.hint = hint
    }
}

public struct QuizOptionPayload: Codable, Sendable {
    public let id: String
    public let text: String
    
    public init(id: String, text: String) {
        self.id = id
        self.text = text
    }
}

public struct FlashcardBlockPayload: Codable, Sendable {
    public let front: String
    public let back: String
    public var tags: String?
    
    public init(front: String, back: String, tags: String? = nil) {
        self.front = front
        self.back = back
        self.tags = tags
    }
}

public struct DataVizBlockPayload: Codable, Sendable {
    public let format: String  // mermaid, chart, table, math
    public let source: String
    public var title: String?
    
    public init(format: String, source: String, title: String? = nil) {
        self.format = format
        self.source = source
        self.title = title
    }
}

public struct MediaBlockPayload: Codable, Sendable {
    public let url: String
    public var alt: String?
    public var caption: String?
    
    public init(url: String, alt: String? = nil, caption: String? = nil) {
        self.url = url
        self.alt = alt
        self.caption = caption
    }
}

public struct ProgressBlockPayload: Codable, Sendable {
    public let completed: Int
    public let total: Int
    public var label: String?
    
    public init(completed: Int, total: Int, label: String? = nil) {
        self.completed = completed
        self.total = total
        self.label = label
    }
}

public struct InteractiveBlockPayload: Codable, Sendable {
    public let items: [InteractiveItem]
    public var title: String?
    
    public init(items: [InteractiveItem], title: String? = nil) {
        self.items = items
        self.title = title
    }
}

public struct InteractiveItem: Codable, Sendable {
    public let label: String
    public let detail: String
    
    public init(label: String, detail: String) {
        self.label = label
        self.detail = detail
    }
}

public struct MasteryMapBlockPayload: Codable, Sendable {
    public let title: String
    public let nodes: [MasteryNodePayload]
    
    public init(title: String, nodes: [MasteryNodePayload]) {
        self.title = title
        self.nodes = nodes
    }
}

public struct MasteryNodePayload: Codable, Sendable, Identifiable {
    public var id: String { nodeId }
    public let nodeId: String
    public let title: String
    public let status: String
    public var masteryLevel: Double?
    
    public init(nodeId: String = UUID().uuidString, title: String, status: String, masteryLevel: Double? = nil) {
        self.nodeId = nodeId
        self.title = title
        self.status = status
        self.masteryLevel = masteryLevel
    }
}

File: android/app/src/main/java/com/lyo/app/data/api/SmartBlock.kt
package com.lyo.app.data.api

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.annotations.SerializedName
import java.lang.reflect.Type
import java.util.UUID

/**
 * Unified block type shared across chat, classroom, and course generation.
 *
 * Mirrors `SmartBlockType` in Sources/Models/SmartBlock.swift (iOS) and
 * `CHAT_BLOCK_TYPES` in web/src/lib/chat-contract.mjs — the three clients
 * decode the backend's SmartBlock envelope (lyo_app/ai/schemas/smart_block.py)
 * independently, so the vocabulary must stay in lockstep across all three.
 */
enum class SmartBlockType(val wireValue: String) {
    TEXT("text"),
    CODE("code"),
    QUIZ("quiz"),
    FLASHCARD("flashcard"),
    DATA_VIZ("dataViz"),
    MEDIA("media"),
    PROGRESS("progress"),
    INTERACTIVE("interactive"),
    MASTERY_MAP("masteryMap"),
    UNKNOWN("unknown"),
    ;

    companion object {
        /** Handles both camelCase ("dataViz") and snake_case ("data_viz") from the backend. */
        fun fromWire(raw: String?): SmartBlockType {
            if (raw == null) return UNKNOWN
            entries.firstOrNull { it.wireValue == raw }?.let { return it }
            val parts = raw.split("_")
            val camelCased = parts.mapIndexed { index, part ->
                if (index == 0) part else part.replaceFirstChar { it.uppercase() }
            }.joinToString("")
            return entries.firstOrNull { it.wireValue == camelCased } ?: UNKNOWN
        }
    }
}

/** Type-safe content container — each block type has its own payload. */
sealed class SmartBlockContent {
    data class Text(val payload: TextBlockPayload) : SmartBlockContent()
    data class Code(val payload: CodeBlockPayload) : SmartBlockContent()
    data class Quiz(val payload: QuizBlockPayload) : SmartBlockContent()
    data class Flashcard(val payload: FlashcardBlockPayload) : SmartBlockContent()
    data class DataViz(val payload: DataVizBlockPayload) : SmartBlockContent()
    data class Media(val payload: MediaBlockPayload) : SmartBlockContent()
    data class Progress(val payload: ProgressBlockPayload) : SmartBlockContent()
    data class Interactive(val payload: InteractiveBlockPayload) : SmartBlockContent()
    data class MasteryMap(val payload: MasteryMapBlockPayload) : SmartBlockContent()
    data class Unknown(val raw: JsonObject) : SmartBlockContent()
}

/** A single renderable content block with schema versioning. */
data class SmartBlock(
    val id: String,
    val schemaVersion: Int,
    val type: SmartBlockType,
    val subtype: String?,
    val content: SmartBlockContent,
    val metadata: Map<String, JsonElement>?,
)

// ── Content Payloads ─────────────────────────────────────────────────────────
// Field-for-field mirrors of Sources/Models/SmartBlock.swift's payload structs.

data class TextBlockPayload(
    val text: String = "",
    val style: String? = null,
)

data class CodeBlockPayload(
    val language: String = "plain",
    val code: String = "",
    @SerializedName("is_runnable") val isRunnable: Boolean? = null,
)

data class QuizOptionPayload(
    val id: String = "",
    val text: String = "",
    /** The misconception this distractor reveals, when the answer is wrong. */
    val reveals: String? = null,
)

data class QuizBlockPayload(
    val question: String = "",
    val options: List<QuizOptionPayload> = emptyList(),
    /**
     * Present on the wire for the *server's* use when grading — never read
     * here to decide correctness client-side. See SmartBlockRenderer's quiz
     * composable, which renders only off the server's CheckAnswerResult.
     */
    @SerializedName("correct_index") val correctIndex: Int = 0,
    val explanation: String? = null,
    val hint: String? = null,
    /** The "just explain it" opt-out option index, when offered. */
    @SerializedName("bailout_index") val bailoutIndex: Int? = null,
)

data class FlashcardBlockPayload(
    val front: String,
    val back: String,
    val tags: String? = null,
)

data class DataVizBlockPayload(
    /** mermaid, chart, table, math, text */
    val format: String,
    val source: String,
    val title: String? = null,
)

data class MediaBlockPayload(
    val url: String,
    val alt: String? = null,
    val caption: String? = null,
)

data class ProgressBlockPayload(
    val completed: Int,
    val total: Int,
    val label: String? = null,
)

data class InteractiveItem(
    val label: String,
    val detail: String,
)

data class InteractiveBlockPayload(
    val items: List<InteractiveItem> = emptyList(),
    val title: String? = null,
)

data class MasteryNodePayload(
    @SerializedName("node_id") val nodeId: String = UUID.randomUUID().toString(),
    val title: String,
    val status: String,
    @SerializedName("mastery_level") val masteryLevel: Double? = null,
)

data class MasteryMapBlockPayload(
    val title: String,
    val nodes: List<MasteryNodePayload> = emptyList(),
)

/**
 * Routes the `content` object to the right typed payload based on the
 * sibling `type` field. Gson has no native sealed-class polymorphism, so
 * this does by hand what `SmartBlock.init(from decoder:)` does on iOS and
 * what the backend's discriminated union does in Python: fail closed to
 * `.Unknown` (never crash, never render raw JSON) rather than throw when a
 * payload doesn't match its declared type.
 *
 * Registered on the shared `ApiClient.gson` instance, so this applies
 * uniformly to Retrofit responses (a reloaded conversation's `blocks`) and
 * the manually-parsed SSE `smart_blocks` event alike.
 */
class SmartBlockDeserializer : JsonDeserializer<SmartBlock> {
    override fun deserialize(
        json: JsonElement,
        typeOfT: Type,
        context: JsonDeserializationContext,
    ): SmartBlock {
        val obj = json.asJsonObject
        val id = obj.get("id")?.takeIf { it.isJsonPrimitive }?.asString ?: UUID.randomUUID().toString()
        val schemaVersion = obj.get("schema_version")?.takeIf { it.isJsonPrimitive }?.asInt
            ?: obj.get("schemaVersion")?.takeIf { it.isJsonPrimitive }?.asInt
            ?: 1
        val type = SmartBlockType.fromWire(obj.get("type")?.takeIf { it.isJsonPrimitive }?.asString)
        val subtype = obj.get("subtype")?.takeIf { it.isJsonPrimitive }?.asString
        val contentElement = obj.get("content")?.takeIf { it.isJsonObject } ?: JsonObject()
        val metadata = obj.get("metadata")?.takeIf { it.isJsonObject }
            ?.asJsonObject?.entrySet()?.associate { it.key to it.value }

        val content = try {
            when (type) {
                SmartBlockType.TEXT ->
                    SmartBlockContent.Text(context.deserialize(contentElement, TextBlockPayload::class.java))
                SmartBlockType.CODE ->
                    SmartBlockContent.Code(context.deserialize(contentElement, CodeBlockPayload::class.java))
                SmartBlockType.QUIZ ->
                    SmartBlockContent.Quiz(context.deserialize(contentElement, QuizBlockPayload::class.java))
                SmartBlockType.FLASHCARD ->
                    SmartBlockContent.Flashcard(context.deserialize(contentElement, FlashcardBlockPayload::class.java))
                SmartBlockType.DATA_VIZ ->
                    SmartBlockContent.DataViz(context.deserialize(contentElement, DataVizBlockPayload::class.java))
                SmartBlockType.MEDIA ->
                    SmartBlockContent.Media(context.deserialize(contentElement, MediaBlockPayload::class.java))
                SmartBlockType.PROGRESS ->
                    SmartBlockContent.Progress(context.deserialize(contentElement, ProgressBlockPayload::class.java))
                SmartBlockType.INTERACTIVE ->
                    SmartBlockContent.Interactive(context.deserialize(contentElement, InteractiveBlockPayload::class.java))
                SmartBlockType.MASTERY_MAP ->
                    SmartBlockContent.MasteryMap(context.deserialize(contentElement, MasteryMapBlockPayload::class.java))
                SmartBlockType.UNKNOWN -> SmartBlockContent.Unknown(contentElement.asJsonObject)
            }
        } catch (e: Exception) {
            SmartBlockContent.Unknown(contentElement.asJsonObject)
        }

        return SmartBlock(
            id = id,
            schemaVersion = schemaVersion,
            type = type,
            subtype = subtype,
            content = content,
            metadata = metadata,
        )
    }
}

// ── Check-grading (mirrors CheckAnswerResult.swift / web's CheckAnswerResult) ─

/**
 * A learner's answer to an in-chat check, sent to the server for grading.
 *
 * Deliberately carries neither the question nor the answer key — the server
 * reads those back from the block it already emitted, so the client cannot
 * assert its own correctness.
 */
data class CheckAnswerRequest(
    @SerializedName("conversation_id") val conversationId: String,
    @SerializedName("block_id") val blockId: String,
    @SerializedName("selected_index") val selectedIndex: Int,
    @SerializedName("time_taken_ms") val timeTakenMs: Int = 0,
    @SerializedName("hint_used") val hintUsed: Boolean = false,
)

/**
 * Server verdict for an answered check. The client renders this — it never
 * computes `correct` itself, and `QuizBlockPayload.correctIndex` is never
 * used to decide it.
 */
data class CheckAnswerResult(
    val correct: Boolean,
    @SerializedName("correct_index") val correctIndex: Int,
    /** Echoed back so the chosen option can be highlighted after a reload. */
    @SerializedName("selected_index") val selectedIndex: Int,
    val explanation: String? = null,
    /** The confusion this particular wrong answer reveals, when known. */
    val misconception: String? = null,
    @SerializedName("bailed_out") val bailedOut: Boolean = false,
    @SerializedName("skill_id") val skillId: String? = null,
    val mastery: Double? = null,
)
